From: Stained Glass OS <ke7oxh@gmail.com>
Subject: [PATCH] msvcp140: Implement the remaining __crt wrappers of kernel32 functions.

msvcp140's remaining cdecl wrappers of kernel32 functions were spec stubs:
__crtCompareStringEx and __crtLCMapStringEx (without the version and sort
handle arguments), __crtGetLocaleInfoEx, __crtGetSystemTimePreciseAsFileTime,
__crtInitOnceExecuteOnce, __crtIsPackagedApp (GetCurrentPackageId),
__crtCreateSymbolicLinkW and __crtGet/SetFileInformationByHandle(Ex).
---
 dlls/msvcp140/msvcp140.spec | 18 ++++-----
 dlls/msvcp90/misc.c         | 89 +++++++++++++++++++++++++++++++++++++++++++++
 2 files changed, 98 insertions(+), 9 deletions(-)

diff --git a/dlls/msvcp140/msvcp140.spec b/dlls/msvcp140/msvcp140.spec
index 1c4621c..9d78661 100644
--- a/dlls/msvcp140/msvcp140.spec
+++ b/dlls/msvcp140/msvcp140.spec
@@ -3756,27 +3756,27 @@
 @ cdecl __crtCloseThreadpoolTimer(ptr) MSVCP__crtCloseThreadpoolTimer
 @ cdecl __crtCloseThreadpoolWait(ptr) MSVCP__crtCloseThreadpoolWait
 @ cdecl __crtCompareStringA(long long str long str long) msvcrt.__crtCompareStringA
-@ stub __crtCompareStringEx
+@ cdecl __crtCompareStringEx(wstr long wstr long wstr long) MSVCP__crtCompareStringEx
 @ cdecl __crtCompareStringW(long long wstr long wstr long) msvcrt.__crtCompareStringW
 @ cdecl __crtCreateEventExW(ptr wstr long long) MSVCP__crtCreateEventExW
 @ cdecl __crtCreateSemaphoreExW(ptr long long wstr long long) MSVCP__crtCreateSemaphoreExW
-@ stub __crtCreateSymbolicLinkW
+@ cdecl __crtCreateSymbolicLinkW(wstr wstr long) MSVCP__crtCreateSymbolicLinkW
 @ cdecl __crtCreateThreadpoolTimer(ptr ptr ptr) MSVCP__crtCreateThreadpoolTimer
 @ cdecl __crtCreateThreadpoolWait(ptr ptr ptr) MSVCP__crtCreateThreadpoolWait
 @ cdecl __crtFlushProcessWriteBuffers() MSVCP__crtFlushProcessWriteBuffers
 @ cdecl __crtFreeLibraryWhenCallbackReturns(ptr ptr) MSVCP__crtFreeLibraryWhenCallbackReturns
 @ cdecl __crtGetCurrentProcessorNumber() MSVCP__crtGetCurrentProcessorNumber
-@ stub __crtGetFileInformationByHandleEx
-@ stub __crtGetLocaleInfoEx
-@ stub __crtGetSystemTimePreciseAsFileTime
+@ cdecl __crtGetFileInformationByHandleEx(long long ptr long) MSVCP__crtGetFileInformationByHandleEx
+@ cdecl __crtGetLocaleInfoEx(wstr long ptr long) MSVCP__crtGetLocaleInfoEx
+@ cdecl __crtGetSystemTimePreciseAsFileTime(ptr) MSVCP__crtGetSystemTimePreciseAsFileTime
 @ cdecl -ret64 __crtGetTickCount64() MSVCP__crtGetTickCount64
-@ stub __crtInitOnceExecuteOnce
+@ cdecl __crtInitOnceExecuteOnce(ptr ptr ptr ptr) MSVCP__crtInitOnceExecuteOnce
 @ cdecl __crtInitializeCriticalSectionEx(ptr long long) MSVCP__crtInitializeCriticalSectionEx
-@ stub __crtIsPackagedApp
+@ cdecl __crtIsPackagedApp() MSVCP__crtIsPackagedApp
 @ cdecl __crtLCMapStringA(long long str long ptr long long long) msvcrt.__crtLCMapStringA
-@ stub __crtLCMapStringEx
+@ cdecl __crtLCMapStringEx(wstr long wstr long ptr long) MSVCP__crtLCMapStringEx
 @ cdecl __crtLCMapStringW(long long wstr long ptr long long long) msvcrt.__crtLCMapStringW
-@ stub __crtSetFileInformationByHandle
+@ cdecl __crtSetFileInformationByHandle(long long ptr long) MSVCP__crtSetFileInformationByHandle
 @ cdecl __crtSetThreadpoolTimer(ptr ptr long long) MSVCP__crtSetThreadpoolTimer
 @ cdecl __crtSetThreadpoolWait(ptr long ptr) MSVCP__crtSetThreadpoolWait
 @ cdecl __crtWaitForThreadpoolTimerCallbacks(ptr long) MSVCP__crtWaitForThreadpoolTimerCallbacks
diff --git a/dlls/msvcp90/misc.c b/dlls/msvcp90/misc.c
index c1d2719..8f2e92a 100644
--- a/dlls/msvcp90/misc.c
+++ b/dlls/msvcp90/misc.c
@@ -23,6 +23,7 @@
 #include <stdlib.h>
 
 #include "msvcp90.h"
+#include "winnls.h"
 
 #include "wine/debug.h"
 
@@ -1717,6 +1718,94 @@ VOID CDECL MSVCP__crtFlushProcessWriteBuffers(void)
     return FlushProcessWriteBuffers();
 }
 
+LONG WINAPI GetCurrentPackageId(UINT32 *, BYTE *);
+
+/*********************************************************************
+ *  __crtCompareStringEx (MSVCP140.@)
+ */
+int CDECL MSVCP__crtCompareStringEx(const WCHAR *locale, DWORD flags, const WCHAR *str1, int len1,
+        const WCHAR *str2, int len2)
+{
+    TRACE("(%s %#lx %s %d %s %d)\n", debugstr_w(locale), flags, debugstr_wn(str1, len1), len1,
+            debugstr_wn(str2, len2), len2);
+    return CompareStringEx(locale, flags, str1, len1, str2, len2, NULL, NULL, 0);
+}
+
+/*********************************************************************
+ *  __crtLCMapStringEx (MSVCP140.@)
+ */
+int CDECL MSVCP__crtLCMapStringEx(const WCHAR *locale, DWORD flags, const WCHAR *src, int srclen,
+        WCHAR *dst, int dstlen)
+{
+    TRACE("(%s %#lx %s %d %p %d)\n", debugstr_w(locale), flags, debugstr_wn(src, srclen), srclen, dst, dstlen);
+    return LCMapStringEx(locale, flags, src, srclen, dst, dstlen, NULL, NULL, 0);
+}
+
+/*********************************************************************
+ *  __crtGetLocaleInfoEx (MSVCP140.@)
+ */
+int CDECL MSVCP__crtGetLocaleInfoEx(const WCHAR *locale, LCTYPE type, WCHAR *data, int len)
+{
+    TRACE("(%s %#lx %p %d)\n", debugstr_w(locale), type, data, len);
+#ifdef SG_MUTANT_CRT_LOCALEINFO_ZERO
+    return 0;
+#endif
+    return GetLocaleInfoEx(locale, type, data, len);
+}
+
+/*********************************************************************
+ *  __crtGetSystemTimePreciseAsFileTime (MSVCP140.@)
+ */
+void CDECL MSVCP__crtGetSystemTimePreciseAsFileTime(FILETIME *ft)
+{
+    GetSystemTimePreciseAsFileTime(ft);
+}
+
+/*********************************************************************
+ *  __crtInitOnceExecuteOnce (MSVCP140.@)
+ */
+BOOL CDECL MSVCP__crtInitOnceExecuteOnce(INIT_ONCE *once, PINIT_ONCE_FN func, void *param, void **context)
+{
+    TRACE("(%p %p %p %p)\n", once, func, param, context);
+    return InitOnceExecuteOnce(once, func, param, context);
+}
+
+/*********************************************************************
+ *  __crtIsPackagedApp (MSVCP140.@)
+ */
+BOOL CDECL MSVCP__crtIsPackagedApp(void)
+{
+    UINT32 len = 0;
+    return GetCurrentPackageId(&len, NULL) != APPMODEL_ERROR_NO_PACKAGE;
+}
+
+/*********************************************************************
+ *  __crtCreateSymbolicLinkW (MSVCP140.@)
+ */
+BOOLEAN CDECL MSVCP__crtCreateSymbolicLinkW(const WCHAR *link, const WCHAR *target, DWORD flags)
+{
+    TRACE("(%s %s %#lx)\n", debugstr_w(link), debugstr_w(target), flags);
+    return CreateSymbolicLinkW(link, target, flags);
+}
+
+/*********************************************************************
+ *  __crtGetFileInformationByHandleEx (MSVCP140.@)
+ */
+BOOL CDECL MSVCP__crtGetFileInformationByHandleEx(HANDLE file, FILE_INFO_BY_HANDLE_CLASS class, void *info, DWORD size)
+{
+    TRACE("(%p %d %p %lu)\n", file, class, info, size);
+    return GetFileInformationByHandleEx(file, class, info, size);
+}
+
+/*********************************************************************
+ *  __crtSetFileInformationByHandle (MSVCP140.@)
+ */
+BOOL CDECL MSVCP__crtSetFileInformationByHandle(HANDLE file, FILE_INFO_BY_HANDLE_CLASS class, void *info, DWORD size)
+{
+    TRACE("(%p %d %p %lu)\n", file, class, info, size);
+    return SetFileInformationByHandle(file, class, info, size);
+}
+
 /*********************************************************************
  *  __crtCreateSemaphoreExW (MSVCP140.@)
  */
